[Serve] feat: make ray.serve.batch concurrent

ArthurBook opened the pull request:
@abrarsheikh 

## Why are these changes needed?
The [current implementation for ray.serve.batch](https://github.com/ray-project/ray/blob/58c081a72aecc33ae31797320ab3a4e17ef02b7f/python/ray/serve/batching.py#L277) executes the batches synchronously. This throttles throughput for asynchronous methods wrapped in `ray.serve.batch`.

This PR introduces a new parameter, `max_concurrent_batches` for `ray.serve.batch`. This enables users to control the concurrency of their `ray.serve.batch` wrapped methods such that `max_concurrent_batches`>=1 concurrent requests are processed by the wrapped method at a time. The parameter defaults to 1, keeping exact backwards compatibility with the previous behavior.

## Related issue number
Closes #53071.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ArthurBook committed:
[Serve] feat: make ray.serve.batch concurrent

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook commented:
Tested with modified repro script from #53071
```python
from typing import List

import asyncio
import time
from ray import serve
from ray.serve.handle import DeploymentHandle


@serve.deployment(max_ongoing_requests=(8 * 3))
class Model:
    @serve.batch(max_batch_size=8, batch_wait_timeout_s=5, max_concurrent_batches=5)
    async def __call__(self, multiple_samples: List[int]) -> List[int]:
        print(f"Processing batch of size {len(multiple_samples)} at {time.strftime('%Y-%m-%d %H:%M:%S')}")
        await asyncio.sleep(1)
        return [i * 2 for i in multiple_samples]


async def main():
    handle: DeploymentHandle = serve.run(Model.bind())
    responses = await asyncio.gather(*map(handle.remote, range(8 * 3)))
    for response in responses:
        print(response)


if __name__ == "__main__":
    asyncio.run(main())
```

```log
underwent 6s ❯ python test.py
/Users/arthurbook/projects/ray/.venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(
2025-05-16 15:16:30,355 INFO worker.py:1879 -- Started a local Ray instance. View the dashboard at http://127.0.0.1:8265
INFO 2025-05-16 15:16:31,549 serve 57512 -- Started Serve in namespace "serve".
(ProxyActor pid=57841) INFO 2025-05-16 15:16:31,529 proxy 127.0.0.1 -- Proxy starting on node 1cd98eb69193e21bfa1f97f9ebaaa80ab7cccb214c8cdaae6dfdf091 (HTTP port: 8000).
(ProxyActor pid=57841) INFO 2025-05-16 15:16:31,543 proxy 127.0.0.1 -- Got updated endpoints: {}.
(ServeController pid=57837) INFO 2025-05-16 15:16:31,576 controller 57837 -- Deploying new version of Deployment(name='Model', app='default') (initial target replicas: 1).
(ProxyActor pid=57841) INFO 2025-05-16 15:16:31,578 proxy 127.0.0.1 -- Got updated endpoints: {Deployment(name='Model', app='default'): EndpointInfo(route='/', app_is_cross_language=False)}.
(ProxyActor pid=57841) INFO 2025-05-16 15:16:31,582 proxy 127.0.0.1 -- Started <ray.serve._private.router.SharedRouterLongPollClient object at 0x12facf610>.
(ServeController pid=57837) INFO 2025-05-16 15:16:31,678 controller 57837 -- Adding 1 replica to Deployment(name='Model', app='default').
INFO 2025-05-16 15:16:32,663 serve 57512 -- Application 'default' is ready at http://127.0.0.1:8000/.
INFO 2025-05-16 15:16:32,667 serve 57512 -- Started <ray.serve._private.router.SharedRouterLongPollClient object at 0x115f1b4f0>.
(ServeReplica:default:Model pid=57848) Processing batch of size 8 at 2025-05-16 15:16:32
(ServeReplica:default:Model pid=57848) Processing batch of size 8 at 2025-05-16 15:16:32
(ServeReplica:default:Model pid=57848) Processing batch of size 8 at 2025-05-16 15:16:32
0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,682 default_Model evz7qqbi 0ea256a4-c5fc-4793-a4ba-c37f4a11fcac -- CALL __call__ OK 1008.2ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,683 default_Model evz7qqbi 25ade65d-5a44-4c97-9b90-4a05d76eca92 -- CALL __call__ OK 1008.0ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,683 default_Model evz7qqbi cc78b021-42ca-4ade-afad-f3e22a90e51e -- CALL __call__ OK 1008.3ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,684 default_Model evz7qqbi d25508a7-0aa5-4d42-8525-73a256a4d07e -- CALL __call__ OK 1009.0ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,684 default_Model evz7qqbi 77915b00-6d39-4dbf-9d11-7ec267f3eb9e -- CALL __call__ OK 1008.4ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,685 default_Model evz7qqbi d6d0483f-adfa-4789-a6fc-5a330334f120 -- CALL __call__ OK 1008.6ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,685 default_Model evz7qqbi 9f1b87f3-5ade-4fd1-9242-45e9f789d5b6 -- CALL __call__ OK 1008.4ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,685 default_Model evz7qqbi 9ac3d690-beec-470d-bd2c-888c7cff2e3a -- CALL __call__ OK 1008.2ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,686 default_Model evz7qqbi 0f1faf1e-1f52-47a3-9df9-f6a4f6f4adb3 -- CALL __call__ OK 1009.0ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,687 default_Model evz7qqbi d96d441a-874d-4cee-b04e-c092db931b00 -- CALL __call__ OK 1008.8ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,687 default_Model evz7qqbi b475538f-91bc-45c0-a050-6e00a15d9c4d -- CALL __call__ OK 1009.0ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,687 default_Model evz7qqbi 4a7ca7d3-cf2b-42c6-809c-d2c46a557ec2 -- CALL __call__ OK 1008.0ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,687 default_Model evz7qqbi bf5cb7b0-eed5-463f-8f5a-eba02dc7c2ba -- CALL __call__ OK 1008.1ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,688 default_Model evz7qqbi 1dc29d40-bbc4-419c-89fb-b386b93540c8 -- CALL __call__ OK 1008.5ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,688 default_Model evz7qqbi 773dad01-bb84-477b-83cc-fe212a5cc857 -- CALL __call__ OK 1008.9ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,688 default_Model evz7qqbi 25ca43d4-56f5-41cc-a01b-8adc5b0cf6da -- CALL __call__ OK 1009.1ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi 739c3daa-512c-4160-8f28-db93a703814d -- CALL __call__ OK 1009.1ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi 650eab61-86e5-4b68-b26d-32ab40c993f1 -- CALL __call__ OK 1008.8ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi 8aaa5e94-7b10-413d-a350-7eac01879cbd -- CALL __call__ OK 1009.0ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi 199418ac-8df7-4d2f-9a64-90bf6a369a26 -- CALL __call__ OK 1009.1ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi fe557276-b7b4-464d-9953-3c754b4db236 -- CALL __call__ OK 1008.7ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi 0c83e1f1-3a6f-412c-a614-5e7143b17f67 -- CALL __call__ OK 1008.7ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,689 default_Model evz7qqbi d0d6fc6f-7305-4fdf-a289-3fa8572dbeb1 -- CALL __call__ OK 1007.2ms
(ServeReplica:default:Model pid=57848) INFO 2025-05-16 15:16:33,690 default_Model evz7qqbi 00d9150b-cbc7-4228-9b0a-3df2afff8594 -- CALL __call__ OK 1007.3ms
```

ArthurBook commented:
Last open question is what to do with [self.curr_iteration_start_time](https://github.com/ray-project/ray/blob/38a1879c58390682a31aa2de545472080b65620f/python/ray/serve/batching.py#L283-L285) since it no longer has the same interpretation.

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -502,6 +530,7 @@ def batch(
     /,
     max_batch_size: int = 10,
     batch_wait_timeout_s: float = 0.0,
+    max_concurrent_batches: int = 1,
```
i am open to considering a change in API.

Is it possible to avoid having this knob? Is it necessary or more of a nice to have?

ArthurBook replied:
I would say it improves user friendliness a lot.

The current default will behave the exactly same way as before (sync), keeping it fully backwards compatible.

Having no concurrency cap would probably cause issues for people less familiar with asynchronous programming, as it would spam their batch methods.

Personally, the jobs I'm running will benefit from having this nob by allowing me to control how many requests I have in the actor at once.

abrarsheikh replied:
That makes sense. What I am wondering is that max_concurrent_batches is a function of (max_batch_size and max_ongoing_request). Interplay of those three quantities maybe confusing. Can we derived max_concurrent_batches from the other two?

ArthurBook replied:
Unfortunately, I don't think so, there are many external things that affect the optimal value for `max_concurrent_batches` such as rate limits on a third party API calls or risk of the event loop clogging within the wrapped method.

abrarsheikh replied:
i see your point

ArthurBook committed:
[Serve] docs: update ray.serve.batch docstring and doc to reflect new parameter

Signed-off-by: Arthur <atte.book@gmail.com>

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -276,61 +259,64 @@ async def _assign_func_results(
     async def _process_batches(self, func: Callable) -> None:
         """Loops infinitely and processes queued request batches."""
-
+        tasks: List[asyncio.Task] = []
```
Use a set instead of a list for efficiency.

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -276,61 +259,64 @@ async def _assign_func_results(
-                )
-
-    async def _process_batch(self, func: Callable) -> None:
+            self.curr_iteration_start_time = time.time()  # TODO: what do we do when there are many ongoing at once?
```
I see that `curr_iteration_start_time` is not used elsewhere in the code. which means it's easier to change what it means. 

I suggest changing `_get_curr_iteration_start_time` to `_get_curr_iteration_start_times` and have it result summary statistics [count, min, max, avg]

ArthurBook replied:
Pushed a suggestion for returning the ongoing start times as a per-task dictionary [here](https://github.com/ray-project/ray/pull/53096/commits/7d365d89f3c1d3c7494e68ca1c717ca1c1730c92)

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -276,61 +259,64 @@ async def _assign_func_results(
-    async def _process_batch(self, func: Callable) -> None:
+            self.curr_iteration_start_time = time.time()  # TODO: what do we do when there are many ongoing at once?
+            batch = await self.wait_for_batch()
+            promise = self._process_batch(func, batch, self.semaphore)
```
why not access `self.semaphore` direct in `_process_batch`?

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -276,61 +259,64 @@ async def _assign_func_results(
+
+    async def _process_batch(self, func: Callable, batch: List[_SingleRequest], semaphore: asyncio.Semaphore) -> None:
         """Processes queued request batch."""
+        async with semaphore:
```
add a comment here explain what this semaphore does.

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -276,61 +259,64 @@ async def _assign_func_results(
+            if task.done():
+                try:
+                    await task
+                except Exception:
+                    logger.exception("_process_batches asyncio task ran into an unexpected exception.")
```
Is it possible for an exception to be thrown here, given we already safeguard against this in `_process_batch`, or is this more of a defense?

ArthurBook replied:
this was also there in the [previous implementation](https://github.com/ray-project/ray/blob/70ec7de3dfe40f2882bca42907ca95ab56d4c5a8/python/ray/serve/batching.py#L281-L287). I think it's just an extra safeguard.

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -276,61 +259,64 @@ async def _assign_func_results(
+            tasks[-1].add_done_callback(lambda task: tasks.remove(task))
+            await self._poll_tasks(tasks)
+
+    async def _process_batch(self, func: Callable, batch: List[_SingleRequest], semaphore: asyncio.Semaphore) -> None:
```
can you confirm that the only change to this function is the introduction of semaphore?

ArthurBook replied:
indeed, nothing but the indentation was changed!

abrarsheikh commented on python/ray/serve/batching.py:
```diff
@@ -113,19 +111,15 @@ def __init__(
+            self._handle_batch_task = self._loop.create_task(self._process_batches(handle_batch_func))
         self._warn_if_max_batch_size_exceeds_max_ongoing_requests()
 
     def _warn_if_max_batch_size_exceeds_max_ongoing_requests(self):
```
this function should now take into account `max_concurrent_batches`?

ArthurBook replied:
could you guide a bit what you mean? I'm not sure I fully understand.

abrarsheikh replied:
```python
if max_ongoing_requests < self.max_batch_size * self. max_concurrent_batches:
    ...
```

ArthurBook committed:
fix: address PR comments

+ Use a set instead of a list for efficiency.

Signed-off-by: Arthur <atte.book@gmail.com>

access self.semaphore directly in _process_batch

Signed-off-by: Arthur <atte.book@gmail.com>

add comment for what the semaphore does is `_process_batch`

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
suggestion: return dictionary of per-ongoing-task start times in `_get_curr_iteration_start_times()`

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
style: lint

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
tests: update `ray.serve.batch` tests to reflect new `_get_curr_iteration_start_times` return type

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
Merge branch 'master' into arthurbook/concurrent-ray.serve.batch

ArthurBook committed:
style: pre-commit

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
fix: remap _get_curr_iteration_start_time causing test fail

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
Merge branch 'master' into arthurbook/concurrent-ray.serve.batch

ArthurBook committed:
fix: update `_warn_if_max_batch_size_exceeds_max_ongoing_requests` to account for concurrency

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
tests: add concurrent batching test

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
Merge branch 'master' into arthurbook/concurrent-ray.serve.batch

abrarsheikh reviewed (approved):
nice, lgtm

ArthurBook commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -34,6 +34,62 @@ async def __call__(self, request):
+def test_concurrent_batching(serve_instance):
+    BATCHES_IN_FLIGHT = 2
+    MAX_BATCH_SIZE = 5
+    BATCH_WAIT_TIMEOUT_S = 1
```
as long as it is a non-zero (by a little margin), it should be fine. The requests are all sent at once so this just makes sure that we gather a full batch before passing to the method. It runs reliably with as small as `BATCH_WAIT_TIMEOUT_S = 0.005`

ArthurBook committed:
tests: address PR comment

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook commented on python/ray/serve/tests/test_batching.py:
```diff
@@ -34,6 +34,62 @@ async def __call__(self, request):
+    result_futures = [handle.remote(i) for i in idxs]
+    result_list = [future.result() for future in result_futures]
+
+    for idx, batches_in_flight, requests_in_flight in result_list:
```
done

ArthurBook committed:
Merge branch 'arthurbook/concurrent-ray.serve.batch' of github.com:ArthurBook/ray into arthurbook/concurrent-ray.serve.batch

ArthurBook committed:
tests: update tests to reflect new `ray.serve.batch` behaviour

Signed-off-by: Arthur <atte.book@gmail.com>

tests: fix warning log message test

Signed-off-by: Arthur <atte.book@gmail.com>

tests: fix test race condition when checking ongoing task times

Signed-off-by: Arthur <atte.book@gmail.com>

tests: try without yield

Signed-off-by: Arthur <atte.book@gmail.com>

fix: return summary statistics instead of dict[Task,float] in
`_get_curr_iteration_start_times`

Signed-off-by: Arthur <atte.book@gmail.com>

ArthurBook committed:
Merge branch 'master' into arthurbook/concurrent-ray.serve.batch

Signed-off-by: Arthur <atte.book@gmail.com>

zcin commented on python/ray/serve/batching.py:
```diff
@@ -276,55 +306,72 @@ async def _assign_func_results(
-            else:
-                func_future = func_future_or_generator
-                await self._assign_func_results(func_future, futures, len(batch))
+    async def _poll_tasks(self) -> None:
+        for task in self.tasks:
+            if task.done():
+                try:
+                    await task
+                except Exception:
+                    logger.exception(
+                        "_process_batches asyncio task ran into an unexpected exception."
+                    )
```
will this code execute first or will the `_handle_completed_task` callback execute first? can we handle exceptions in the callback instead?

ArthurBook replied:
Hi @zcin! Thanks for taking a look at the PR!

I tried to minimize the diff and kept two try except statements (like before):
- one inside the task execution [here in my implementation](https://github.com/ray-project/ray/blob/0876c2c271aef3b0a6a1e3e7934b9ceba460da8f/python/ray/serve/batching.py#L334), from [here in master](https://github.com/ray-project/ray/blob/bc0a483e624b6710735f7defe51cc77ef492467c/python/ray/serve/batching.py#L305-L327)
- one where the tasks are polled, [here in my implementation](https://github.com/ray-project/ray/blob/0876c2c271aef3b0a6a1e3e7934b9ceba460da8f/python/ray/serve/batching.py#L365-L370) from [here in master](https://github.com/ray-project/ray/blob/bc0a483e624b6710735f7defe51cc77ef492467c/python/ray/serve/batching.py#L280-L287)

`asyncio.Future` [callbacks trigger when the promise is marked as _done_](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback), i.e. either a [result is set](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback), or [an exception is raised](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback).

However, because there is a try-catch already inside the task execution, I don't think we can ever have an exception raised to the callback... We discussed this with @abrarsheikh [here](https://github.com/ray-project/ray/pull/53096#discussion_r2098636539). I think we can safely remove the try-catch where the tasks are polled. WDYT?

zcin reviewed (commented):
Overall LGTM! one comment about handling batch completion

ArthurBook committed:
merge master

ArthurBook committed:
Merge branch 'master' into arthurbook/concurrent-ray.serve.batch

ArthurBook committed:
Merge branch 'master' into arthurbook/concurrent-ray.serve.batch

ArthurBook committed:
fix: move exception logging into callback

Signed-off-by: Arthur <atte.book@gmail.com>
